File: stale-negative-cache-nxdomain/python/stale_negative_cache.py
"""Detect a stale negative-cache NXDOMAIN across public resolvers.
Safe to run on a schedule. Only reads, unless you also lower the zone's
negative-cache TTL, which is guarded by DRY_RUN.
"""
import os
import logging

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("stale_negative_cache")


def stale_negative_cache_report(soa_minimum: int, soa_ttl_seen: int,
                                 resolver_results: dict, authoritative_has_record: bool) -> dict:
    """
    Pure decision logic, no I/O.
    - soa_minimum: the SOA MINIMUM field (negative-cache TTL) in seconds, from RFC 2308.
    - soa_ttl_seen: the TTL value seen on the SOA record inside an NXDOMAIN authority section
      (countdown remaining).
    - resolver_results: {resolver_ip: (rcode, sample_ttl)} e.g.
      {"8.8.8.8": ("NXDOMAIN", 2143), "1.1.1.1": ("NOERROR", 299)}.
    - authoritative_has_record: True if the authoritative server currently answers NOERROR
      for the name.
    Returns: {"stale_resolvers": [...], "eta_seconds": {...}, "is_stale_negative_cache": bool}
    """
    stale = []
    eta = {}
    for resolver, (rcode, ttl) in resolver_results.items():
        if authoritative_has_record and rcode == "NXDOMAIN":
            stale.append(resolver)
            eta[resolver] = max(ttl, 0)  # seconds remaining until this resolver's entry expires
    return {
        "stale_resolvers": stale,
        "eta_seconds": eta,
        "is_stale_negative_cache": authoritative_has_record and len(stale) > 0,
        "max_wait_seconds": max(eta.values()) if eta else 0,
    }


def run():
    # Imported lazily so the pure function above can be tested with no
    # network libraries installed at all.
    import dns.resolver
    import dns.rdatatype
    import requests

    domain = os.environ["DNS_DOMAIN"]
    zone = os.environ.get("DNS_ZONE", domain)
    public_resolvers = os.environ.get("PUBLIC_RESOLVERS", "1.1.1.1,8.8.8.8,9.9.9.9").split(",")
    dry_run = os.environ.get("DRY_RUN", "true").lower() == "true"

    zone_id = os.environ.get("CLOUDFLARE_ZONE_ID")
    api_token = os.environ.get("CLOUDFLARE_API_TOKEN")

    # Confirm the record at the authoritative server first, bypassing recursion.
    ns_answer = dns.resolver.resolve(zone, "NS")
    nameservers = [str(r.target).rstrip(".") for r in ns_answer]
    ns_ip = dns.resolver.resolve(nameservers[0], "A")[0].address

    auth_resolver = dns.resolver.Resolver()
    auth_resolver.nameservers = [ns_ip]

    authoritative_has_record = True
    try:
        auth_resolver.resolve(domain, "A")
    except dns.resolver.NXDOMAIN:
        authoritative_has_record = False
    except dns.resolver.NoAnswer:
        authoritative_has_record = False

    # Read the zone's SOA MINIMUM (negative-cache TTL, per RFC 2308).
    soa = dns.resolver.resolve(zone, "SOA")[0]
    soa_minimum = soa.minimum

    # Query each public resolver directly and record rcode plus the SOA TTL
    # it shows in the authority section, if any.
    resolver_results = {}
    for ip in public_resolvers:
        r = dns.resolver.Resolver()
        r.nameservers = [ip.strip()]
        try:
            r.resolve(domain, "A")
            resolver_results[ip.strip()] = ("NOERROR", 0)
        except dns.resolver.NXDOMAIN as exc:
            ttl = 0
            response = exc.response() if hasattr(exc, "response") else None
            for rrset in (response.authority if response else []):
                if rrset.rdtype == dns.rdatatype.SOA:
                    ttl = rrset.ttl
            resolver_results[ip.strip()] = ("NXDOMAIN", ttl)
        except dns.resolver.NoAnswer:
            resolver_results[ip.strip()] = ("NOERROR", 0)

    report = stale_negative_cache_report(soa_minimum, 0, resolver_results, authoritative_has_record)
    log.info("Report for %s: %s", domain, report)

    if not report["is_stale_negative_cache"]:
        log.info("No stale negative cache detected.")
        return

    log.info("Stale on %s. Longest wait about %d seconds.",
              ", ".join(report["stale_resolvers"]), report["max_wait_seconds"])

    if not zone_id or not api_token:
        log.info("No Cloudflare credentials set, skipping negative-cache TTL check.")
        return

    settings = requests.get(
        f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_settings",
        headers={"Authorization": f"Bearer {api_token}"},
        timeout=30,
    )
    settings.raise_for_status()
    log.info("Current zone DNS settings: %s", settings.json().get("result"))

    if dry_run:
        log.info("Dry run: would lower the zone's negative-cache TTL to speed up future recovery.")
        return

    resp = requests.patch(
        f"https://api.cloudflare.com/client/v4/zones/{zone_id}/dns_settings",
        headers={"Authorization": f"Bearer {api_token}", "Content-Type": "application/json"},
        json={"ns_ttl": 300},
        timeout=30,
    )
    resp.raise_for_status()
    log.info("Lowered the zone's negative-cache TTL to 300 seconds for next time.")


if __name__ == "__main__":
    run()

File: stale-negative-cache-nxdomain/node/stale-negative-cache.js
/**
 * Detect a stale negative-cache NXDOMAIN across public resolvers.
 * Safe to run on a schedule. Only reads, unless you also lower the zone's
 * negative-cache TTL, which is guarded by DRY_RUN.
 */
import { pathToFileURL } from "node:url";

export function staleNegativeCacheReport(soaMinimum, soaTtlSeen, resolverResults, authoritativeHasRecord) {
  // Pure decision logic, no I/O.
  // resolverResults: { [resolverIp]: [rcode, sampleTtl] }
  const stale = [];
  const eta = {};
  for (const [resolver, [rcode, ttl]] of Object.entries(resolverResults)) {
    if (authoritativeHasRecord && rcode === "NXDOMAIN") {
      stale.push(resolver);
      eta[resolver] = Math.max(ttl, 0); // seconds remaining until this resolver's entry expires
    }
  }
  const etaValues = Object.values(eta);
  return {
    staleResolvers: stale,
    etaSeconds: eta,
    isStaleNegativeCache: authoritativeHasRecord && stale.length > 0,
    maxWaitSeconds: etaValues.length ? Math.max(...etaValues) : 0,
  };
}

export async function run() {
  // Imported lazily so the pure function above can be tested with no
  // network modules touched at all.
  const dns = await import("node:dns");
  const resolvePromises = dns.promises;

  const domain = process.env.DNS_DOMAIN;
  const zone = process.env.DNS_ZONE || domain;
  const publicResolvers = (process.env.PUBLIC_RESOLVERS || "1.1.1.1,8.8.8.8,9.9.9.9").split(",");
  const dryRun = (process.env.DRY_RUN || "true").toLowerCase() === "true";

  const zoneId = process.env.CLOUDFLARE_ZONE_ID;
  const apiToken = process.env.CLOUDFLARE_API_TOKEN;

  // Confirm the record at the authoritative server first, bypassing recursion.
  const nameservers = await resolvePromises.resolveNs(zone);
  const nsIp = (await resolvePromises.resolve4(nameservers[0]))[0];

  const authResolver = new dns.promises.Resolver();
  authResolver.setServers([nsIp]);

  let authoritativeHasRecord = true;
  try {
    await authResolver.resolve4(domain);
  } catch (err) {
    if (err.code === "ENOTFOUND" || err.code === "ENODATA") {
      authoritativeHasRecord = false;
    } else {
      throw err;
    }
  }

  // Read the zone's SOA MINIMUM (negative-cache TTL). Node's resolveSoa
  // exposes it as minttl.
  const soa = await resolvePromises.resolveSoa(zone);
  const soaMinimum = soa.minttl;

  // Query each public resolver directly and record rcode plus a sample TTL.
  const resolverResults = {};
  for (const ip of publicResolvers) {
    const r = new dns.promises.Resolver();
    r.setServers([ip.trim()]);
    try {
      await r.resolve4(domain);
      resolverResults[ip.trim()] = ["NOERROR", 0];
    } catch (err) {
      if (err.code === "ENOTFOUND") {
        // node:dns does not expose the authority-section SOA TTL directly,
        // so a real implementation would parse the raw response; here we
        // record the negative hit with an unknown (0) sample TTL.
        resolverResults[ip.trim()] = ["NXDOMAIN", 0];
      } else if (err.code === "ENODATA") {
        resolverResults[ip.trim()] = ["NOERROR", 0];
      } else {
        throw err;
      }
    }
  }

  const report = staleNegativeCacheReport(soaMinimum, 0, resolverResults, authoritativeHasRecord);
  console.log(`Report for ${domain}:`, report);

  if (!report.isStaleNegativeCache) {
    console.log("No stale negative cache detected.");
    return;
  }

  console.log(`Stale on ${report.staleResolvers.join(", ")}. Longest wait about ${report.maxWaitSeconds} seconds.`);

  if (!zoneId || !apiToken) {
    console.log("No Cloudflare credentials set, skipping negative-cache TTL check.");
    return;
  }

  const settingsRes = await fetch(`https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_settings`, {
    headers: { Authorization: `Bearer ${apiToken}` },
  });
  if (!settingsRes.ok) throw new Error(`Cloudflare API returned ${settingsRes.status}`);
  const settings = await settingsRes.json();
  console.log("Current zone DNS settings:", settings.result);

  if (dryRun) {
    console.log("Dry run: would lower the zone's negative-cache TTL to speed up future recovery.");
    return;
  }

  const patchRes = await fetch(`https://api.cloudflare.com/client/v4/zones/${zoneId}/dns_settings`, {
    method: "PATCH",
    headers: { Authorization: `Bearer ${apiToken}`, "Content-Type": "application/json" },
    body: JSON.stringify({ ns_ttl: 300 }),
  });
  if (!patchRes.ok) throw new Error(`Cloudflare API returned ${patchRes.status}`);
  console.log("Lowered the zone's negative-cache TTL to 300 seconds for next time.");
}

if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
